## Import Statements

In [1]:
# Import Statements
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import chart_studio.plotly as py
import re
import geopy
from geopy.distance import geodesic

from IPython.display import display, HTML

### Read in other2142_runways.csv

`This .csv file contains the 599 remarks that reference runways (of the 2142 OTHER Standard Format Remarks).`

In [2]:
# Read in Files
rwy599 = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/other2142_runways.csv')
rwy599

index                                            REMARKS  \
0        0  Aircraft observed a UAS 100 feet above the Air...   
1        1  Aircraft observed an UAS at 200 feet above the...   
2        2  Aircraft observed a UAS off his right side at ...   
3        3  Aircraft observed a red and white UAS at 1,000...   
4        4  Aircraft observed a black UAS at 1,000 feet wh...   
..     ...                                                ...   
594    596  Aircraft reported a fixed wing UAS pass within...   
595    597  Aircraft reported a grey quad-copter UAS off t...   
596    598  Aircraft reported a black quad-copter UAS off ...   
597    599  Aircraft reported a NMAC with a UAS while on a...   
598    600  VFR Aircraft reported a near mid-air with a UA...   

     UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0               ['RWY30']         NaN   NaN         NaN     ['2 NM']  
1                ['RWY8']         NaN   NaN         NaN           []  
2               ['RWY23']         NaN   NaN         NaN           []  
3    ['RWY18L', 'RWY18C']         NaN   NaN         NaN           []  
4             ['RWY 32R']         PAM   PAM     RWY 32R     ['3 NM']  
..                    ...         ...   ...         ...          ...  
594            ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']  
595           ['RWY 27L']         SFB   SFB     RWY 27L           []  
596            ['RWY 3R']         NaN   NaN         NaN           []  
597             ['RWY05']         NaN   NaN         NaN     ['6 NM']  
598           ['RWY 28R']         BOI   BOI     RWY 28R           []  

[599 rows x 7 columns]

#### We first explore the ***102*** records for which there is NO NULL (or blank) INFORMATION

`These records will have useable information in all 6 columns of interest in the rwy599 dataframe`

In [3]:
# First we need to replace all [] in UAS_Distance field with NaN
rwy599_df = rwy599.mask(rwy599.applymap(str).eq('[]'))
rwy599_df

index                                            REMARKS  \
0        0  Aircraft observed a UAS 100 feet above the Air...   
1        1  Aircraft observed an UAS at 200 feet above the...   
2        2  Aircraft observed a UAS off his right side at ...   
3        3  Aircraft observed a red and white UAS at 1,000...   
4        4  Aircraft observed a black UAS at 1,000 feet wh...   
..     ...                                                ...   
594    596  Aircraft reported a fixed wing UAS pass within...   
595    597  Aircraft reported a grey quad-copter UAS off t...   
596    598  Aircraft reported a black quad-copter UAS off ...   
597    599  Aircraft reported a NMAC with a UAS while on a...   
598    600  VFR Aircraft reported a near mid-air with a UA...   

     UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0               ['RWY30']         NaN   NaN         NaN     ['2 NM']  
1                ['RWY8']         NaN   NaN         NaN          NaN  
2               ['RWY23']         NaN   NaN         NaN          NaN  
3    ['RWY18L', 'RWY18C']         NaN   NaN         NaN          NaN  
4             ['RWY 32R']         PAM   PAM     RWY 32R     ['3 NM']  
..                    ...         ...   ...         ...          ...  
594            ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']  
595           ['RWY 27L']         SFB   SFB     RWY 27L          NaN  
596            ['RWY 3R']         NaN   NaN         NaN          NaN  
597             ['RWY05']         NaN   NaN         NaN     ['6 NM']  
598           ['RWY 28R']         BOI   BOI     RWY 28R          NaN  

[599 rows x 7 columns]

In [4]:
# Drop all NaN from rwy599_df
rwy599_df1 = rwy599_df.dropna().reset_index(drop=True)
rwy599_df1

index                                            REMARKS  \
0        4  Aircraft observed a black UAS at 1,000 feet wh...   
1       85  Aircraft observed a UAS above while W bound at...   
2      109  Aircraft observed a UAS at the same altitude w...   
3      124  Aircraft observed a red UAS off the right side...   
4      145  Aircraft observed a small red UAS E bound off ...   
..     ...                                                ...   
97     516  Aircraft reported a UAS passing 100 feet below...   
98     588  Aircraft reported a UAS off the right side, ap...   
99     591  Aircraft reported a grey, basketball size, UAS...   
100    594  Aircraft reported a UAS off the left side whil...   
101    596  Aircraft reported a fixed wing UAS pass within...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0            ['RWY 32R']         PAM   PAM     RWY 32R     ['3 NM']  
1            ['RWY 18R']         DFW   DFW     RWY 18R     ['5 NM']  
2            ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']  
3            ['RWY 16R']         SEA   SEA     RWY 16R     ['4 NM']  
4             ['RWY 15']         ADS   ADS      RWY 15     ['3 NM']  
..                   ...         ...   ...         ...          ...  
97            ['RWY 27']         TLH   TLH      RWY 27     ['2 NM']  
98           ['RWY 25L']         LAX   LAX     RWY 25L     ['3 NM']  
99           ['RWY 25L']         LAX   LAX     RWY 25L    ['10 NM']  
100          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']  
101           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']  

[102 rows x 7 columns]

In [5]:
# CHECK THAT THERE ARE NO DUPLICATE REMARKS
#rwy599_df1.drop_duplicates(subset=['REMARKS']) 

# THERE ARE STILL 102 ROWS SO WE ARE GOOD TO KEEP GOING

#### Import airports_runways_linked.csv dataset 

`Contains the cleaned and linked airports_cleaned.csv and rwy.csv supplemental datasets`

In [6]:
air_rwy = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/airports_runways_linked.csv')
#air_rwy

#### Merge other_rwy_totalinfo.csv and airports_runways_linked.csv

`Merge the two datasets on the 'IDENT' field.`

In [7]:
rwy_linked = pd.merge(rwy599_df1, air_rwy, on = 'IDENT', how = 'left')
rwy_linked.rename(columns = {'properties.DESIGNATOR': 'DESIGNATOR'}, inplace = True)

In [8]:
rwy_linked

index                                            REMARKS  \
0        4  Aircraft observed a black UAS at 1,000 feet wh...   
1        4  Aircraft observed a black UAS at 1,000 feet wh...   
2       85  Aircraft observed a UAS above while W bound at...   
3       85  Aircraft observed a UAS above while W bound at...   
4       85  Aircraft observed a UAS above while W bound at...   
..     ...                                                ...   
324    594  Aircraft reported a UAS off the left side whil...   
325    594  Aircraft reported a UAS off the left side whil...   
326    594  Aircraft reported a UAS off the left side whil...   
327    596  Aircraft reported a fixed wing UAS pass within...   
328    596  Aircraft reported a fixed wing UAS pass within...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  \
0            ['RWY 32R']         PAM   PAM     RWY 32R     ['3 NM']   
1            ['RWY 32R']         PAM   PAM     RWY 32R     ['3 NM']   
2            ['RWY 18R']         DFW   DFW     RWY 18R     ['5 NM']   
3            ['RWY 18R']         DFW   DFW     RWY 18R     ['5 NM']   
4            ['RWY 18R']         DFW   DFW     RWY 18R     ['5 NM']   
..                   ...         ...   ...         ...          ...   
324          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
325          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
326          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
327           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']   
328           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']   

                     properties.GLOBAL_ID  \
0    8FAB3563-C164-4D2A-A7D7-4CEDC40E1B5D   
1    EC5EA23A-FB43-4AF4-9F3A-6B72C1495F0C   
2    D60C0EEA-7524-4790-A294-CE2112A0E2FB   
3    A4F0B104-48EA-445F-8DAB-850EE0FCBA8C   
4    A088B20F-FBE9-4C4F-9937-F7ABADEE30CC   
..                                    ...   
324  91541445-E5CF-40C1-8E8F-3E759B9CA482   
325  D1A26CC8-C4DA-48CF-A0B0-7FD883EBA2FA   
326  B270C60D-5C0C-41EA-8410-5DA7D520B4E3   
327  B2D0E0C7-E412-458E-B683-52C4DD4E4EA9   
328  FFE4D41C-A5CA-4072-BCE0-59A70192F6D3   

                    properties.AIRPORT_ID DESIGNATOR  Runway.Latitude  \
0    9907F7DB-7C4F-4CB6-98FA-F0462EE27DE9    14L/32R       -85.570825   
1    9907F7DB-7C4F-4CB6-98FA-F0462EE27DE9    14R/32L       -85.579142   
2    B371E693-B91F-40ED-ABB1-628EFE9D3111    13L/31R       -97.013278   
3    B371E693-B91F-40ED-ABB1-628EFE9D3111    13R/31L       -97.075234   
4    B371E693-B91F-40ED-ABB1-628EFE9D3111    17C/35C       -97.026110   
..                                    ...        ...              ...   
324  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4    05L/23R       -78.792358   
325  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4    05R/23L       -78.790127   
326  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4      14/32       -78.777558   
327  D339C530-1D36-4587-8130-9E9938BF474D      01/19       -81.200005   
328  D339C530-1D36-4587-8130-9E9938BF474D      10/28       -81.206715   

     Runway.Longitude                       NAME   LATITUDE  LONGITUDE ICAO_ID  
0           30.068523                Tyndall AFB  30.069047 -85.574981    KPAM  
1           30.069627                Tyndall AFB  30.069047 -85.574981    KPAM  
2           32.905492     Dallas-Fort Worth Intl  32.897233 -97.037695    KDFW  
3           32.901832     Dallas-Fort Worth Intl  32.897233 -97.037695    KDFW  
4           32.900981     Dallas-Fort Worth Intl  32.897233 -97.037695    KDFW  
..                ...                        ...        ...        ...     ...  
324         35.882177        Raleigh-Durham Intl  35.877639 -78.787472    KRDU  
325         35.870421        Raleigh-Durham Intl  35.877639 -78.787472    KRDU  
326         35.870885        Raleigh-Durham Intl  35.877639 -78.787472    KRDU  
327         32.124275  Savannah/Hilton Head Intl  32.127583 -81.202139    KSAV  
328         32.128607  Savannah/Hilton Head Intl  32.127583 -8

#### Extract Runway Designator from RWYLOCATION field

`We want to be able to obtain the correct latitude/longitude coordinates for the appropriate runway mentioned in the remark.`

In [9]:
rwydesignator = []

for i in range(len(rwy_linked)):
    rwy_split = rwy_linked['RWYLOCATION'][i].split(' ')
    rwydesignator.append(rwy_split[1])

# Add split column with RWY designators to dataframe as new field
rwy_linked['RWYDESIGNATOR'] = rwydesignator

In [10]:
#rwy_linked

#### Check if RWY designator found in REMARK matches one of the RWY designators from the rwy.csv dataset

`If YES: RWYinREMARK field is 'True'`

`If NO: RWYinREMARK field is 'False'`
***
***NOTE:*** Since we linked on the 'IDENT' field, there are repeated remarks as a result of there being multiple runways at
each airport. Thus, we are only interested in the records for which the 'RWYinREMARK' field is 'True'.

In [11]:
rwy_linked['RWYinREMARK'] = rwy_linked.apply(lambda x: str(x.RWYDESIGNATOR) in str(x.DESIGNATOR), axis=1)

# Only keep records for which the RWY Designator is present in the DESIGNATOR field 
rwy95 = rwy_linked[rwy_linked['RWYinREMARK'] == True]
rwy95

index                                            REMARKS  \
0        4  Aircraft observed a black UAS at 1,000 feet wh...   
8       85  Aircraft observed a UAS above while W bound at...   
12     109  Aircraft observed a UAS at the same altitude w...   
15     124  Aircraft observed a red UAS off the right side...   
16     145  Aircraft observed a small red UAS E bound off ...   
..     ...                                                ...   
314    516  Aircraft reported a UAS passing 100 feet below...   
319    588  Aircraft reported a UAS off the right side, ap...   
323    591  Aircraft reported a grey, basketball size, UAS...   
324    594  Aircraft reported a UAS off the left side whil...   
328    596  Aircraft reported a fixed wing UAS pass within...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  \
0            ['RWY 32R']         PAM   PAM     RWY 32R     ['3 NM']   
8            ['RWY 18R']         DFW   DFW     RWY 18R     ['5 NM']   
12           ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']   
15           ['RWY 16R']         SEA   SEA     RWY 16R     ['4 NM']   
16            ['RWY 15']         ADS   ADS      RWY 15     ['3 NM']   
..                   ...         ...   ...         ...          ...   
314           ['RWY 27']         TLH   TLH      RWY 27     ['2 NM']   
319          ['RWY 25L']         LAX   LAX     RWY 25L     ['3 NM']   
323          ['RWY 25L']         LAX   LAX     RWY 25L    ['10 NM']   
324          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
328           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']   

                     properties.GLOBAL_ID  \
0    8FAB3563-C164-4D2A-A7D7-4CEDC40E1B5D   
8    ED9E6106-EB1D-401C-BCAE-E7A73A60A41C   
12   DC9631DF-1821-449A-AD41-8EA0B8B40375   
15   D8DB5055-844E-44D9-A4DA-C7E399892BE1   
16   202E6282-A103-44B0-ACCE-3D984B66D091   
..                                    ...   
314  6AFB0072-D2C9-4832-9B3A-9BA008C1E9F9   
319  8ED22315-4327-432F-9C74-65E1A06D97A5   
323  8ED22315-4327-432F-9C74-65E1A06D97A5   
324  91541445-E5CF-40C1-8E8F-3E759B9CA482   
328  FFE4D41C-A5CA-4072-BCE0-59A70192F6D3   

                    properties.AIRPORT_ID DESIGNATOR  Runway.Latitude  \
0    9907F7DB-7C4F-4CB6-98FA-F0462EE27DE9    14L/32R       -85.570825   
8    B371E693-B91F-40ED-ABB1-628EFE9D3111    18R/36L       -97.054815   
12   DA96A371-EB42-4312-8EAF-FA237DE1E404    18R/36L       -89.987278   
15   5338FBAA-B183-4BC0-B3F1-27ABF6D1AD7C    16R/34L      -122.317932   
16   3646423B-964B-466A-9CE9-3CB43457F772      15/33       -96.837304   
..                                    ...        ...              ...   
314  265A7B00-46FF-4744-8673-499589FFA2BF      09/27       -84.346397   
319  2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
323  2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
324  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4    05L/23R       -78.792358   
328  D339C530-1D36-4587-8130-9E9938BF474D      10/28       -81.206715   

     Runway.Longitude                       NAME   LATITUDE   LONGITUDE  \
0           30.068523                Tyndall AFB  30.069047  -85.574981   
8           32.893723     Dallas-Fort Worth Intl  32.897233  -97.037695   
12          35.039251               Memphis Intl  35.042411  -89.976679   
15          47.449860        Seattle-Tacoma Intl  47.449889 -122.311778   
16          32.970408                    Addison  32.968556  -96.836444   
..                ...                        ...        ...         ...   
314         30.391355           Tallahassee Intl  30.396754  -84.350869   
319         33.935085           Los Angeles Intl  33.942496 -118.408049   
323         33.935085           Los Angeles Intl  33.942496 -118.408049   
324         35.882177        Raleigh-Durham Intl  35.877639  -78.787472   
328         32.128607  Savannah/Hilton Head Intl  32.127583  -81.202139   

    ICAO_ID RWYDESIGNATOR  RWYinREMARK  
0      KPAM

#### Import airports_cleaned_declination.csv dataset 

`Contains the cleaned and linked airports_cleaned.csv and declination.csv supplemental datasets`

In [12]:
air_dec = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/airports_cleaned_declination.csv')
air_dec

Column1 IDENT                      NAME   LATITUDE   LONGITUDE ICAO_ID  \
0            0  6TS0                      TRUE  34.089220 -101.740461     NaN   
1            1  WY47                      TRUE  42.853576 -106.336415     NaN   
2            2  WA08              Zwainz Farms  47.713439 -117.919989     NaN   
3            3  9IN7              Zupancic Fld  39.354722  -86.306944     NaN   
4            4  1TE4                     Zuehl  29.495509  -98.157787     NaN   
...        ...   ...                       ...        ...         ...     ...   
19889    19889  55XA  1101 East Blvd Deer Park  29.702772  -95.101639     NaN   
19890    19890  WA54  1001 Fourth Avenue Plaza  47.606765 -122.333737     NaN   
19891    19891  FA10               1000 Museum  25.784228  -80.190017     NaN   
19892    19892  TN41             100 Aker Wood  35.772778  -84.765278     NaN   
19893    19893  IN52                       500  39.782821  -86.233322     NaN   

       DECLINATION  
0          5.74941  
1          8.90491  
2         14.44256  
3         -4.69381  
4          3.68769  
...            ...  
19889      1.95632  
19890     15.44604  
19891     -6.97982  
19892     -5.33458  
19893     -4.80367  

[19894 rows x 7 columns]

#### Merge air_dec with rwy95

`Merge the two datasets on the 'IDENT' field.`

`air_dec --> Contains the cleaned airports supplemental dataset with the Declination value at the airport`

`rwy95 --> Dataframe containing 95 records referencing Runways that contain all relevant information (from the 599 OTHER records)`

In [13]:
rwy95_dec = pd.merge(rwy95, air_dec, on = 'IDENT', how = 'left')

# Only keep necessary columns/fields of interest
rwy95_decDF = pd.DataFrame(rwy95_dec[['REMARKS', 'UASLOCATION', 'IDENT', 'RWYLOCATION', 'UAS_Distance', 'RWYDESIGNATOR', 'Runway.Latitude', 'Runway.Longitude', 'DECLINATION']])

In [14]:
rwy95_decDF

REMARKS UASLOCATION IDENT  \
0   Aircraft observed a black UAS at 1,000 feet wh...         PAM   PAM   
1   Aircraft observed a UAS above while W bound at...         DFW   DFW   
2   Aircraft observed a UAS at the same altitude w...         MEM   MEM   
3   Aircraft observed a red UAS off the right side...         SEA   SEA   
4   Aircraft observed a small red UAS E bound off ...         ADS   ADS   
..                                                ...         ...   ...   
90  Aircraft reported a UAS passing 100 feet below...         TLH   TLH   
91  Aircraft reported a UAS off the right side, ap...         LAX   LAX   
92  Aircraft reported a grey, basketball size, UAS...         LAX   LAX   
93  Aircraft reported a UAS off the left side whil...         RDU   RDU   
94  Aircraft reported a fixed wing UAS pass within...         SAV   SAV   

   RWYLOCATION UAS_Distance RWYDESIGNATOR  Runway.Latitude  Runway.Longitude  \
0      RWY 32R     ['3 NM']           32R       -85.570825         30.068523   
1      RWY 18R     ['5 NM']           18R       -97.054815         32.893723   
2      RWY 36L     ['6 NM']           36L       -89.987278         35.039251   
3      RWY 16R     ['4 NM']           16R      -122.317932         47.449860   
4       RWY 15     ['3 NM']            15       -96.837304         32.970408   
..         ...          ...           ...              ...               ...   
90      RWY 27     ['2 NM']            27       -84.346397         30.391355   
91     RWY 25L     ['3 NM']           25L      -118.404501         33.935085   
92     RWY 25L    ['10 NM']           25L      -118.404501         33.935085   
93     RWY 23R     ['3 NM']           23R       -78.792358         35.882177   
94      RWY 10     ['1 NM']            10       -81.206715         32.128607   

    DECLINATION  
0      -4.13012  
1       3.06900  
2      -1.62738  
3      15.40471  
4       2.94849  
..          ...  
90     -4.96630  
91     11.71212  
92     11.71212  
93     -9.22834  
94     -7.20523  

[95 rows x 9 columns]

### Calculate Bearing Information for Runways

`To do this, you take the number portion of the Runway Designator (in the 'RWYDESIGNATOR' field) and multiply it by 10 (effectively adding a 0 to the end) and add the DECLINATION to this number`

***Example:*** *RWY 32R has designator 32R which indicates a bearing of 320 degrees*

In [15]:
designator = rwy95_decDF['RWYDESIGNATOR']
designator_bearing = []

for i in range(len(rwy95_decDF)):
    designator_bearing.append(int(re.sub("\D","",designator[i]))*10 + rwy95_decDF['DECLINATION'][i])

rwy95_decDF['RWY_BEARING'] = designator_bearing

In [16]:
rwy95_decDF

REMARKS UASLOCATION IDENT  \
0   Aircraft observed a black UAS at 1,000 feet wh...         PAM   PAM   
1   Aircraft observed a UAS above while W bound at...         DFW   DFW   
2   Aircraft observed a UAS at the same altitude w...         MEM   MEM   
3   Aircraft observed a red UAS off the right side...         SEA   SEA   
4   Aircraft observed a small red UAS E bound off ...         ADS   ADS   
..                                                ...         ...   ...   
90  Aircraft reported a UAS passing 100 feet below...         TLH   TLH   
91  Aircraft reported a UAS off the right side, ap...         LAX   LAX   
92  Aircraft reported a grey, basketball size, UAS...         LAX   LAX   
93  Aircraft reported a UAS off the left side whil...         RDU   RDU   
94  Aircraft reported a fixed wing UAS pass within...         SAV   SAV   

   RWYLOCATION UAS_Distance RWYDESIGNATOR  Runway.Latitude  Runway.Longitude  \
0      RWY 32R     ['3 NM']           32R       -85.570825         30.068523   
1      RWY 18R     ['5 NM']           18R       -97.054815         32.893723   
2      RWY 36L     ['6 NM']           36L       -89.987278         35.039251   
3      RWY 16R     ['4 NM']           16R      -122.317932         47.449860   
4       RWY 15     ['3 NM']            15       -96.837304         32.970408   
..         ...          ...           ...              ...               ...   
90      RWY 27     ['2 NM']            27       -84.346397         30.391355   
91     RWY 25L     ['3 NM']           25L      -118.404501         33.935085   
92     RWY 25L    ['10 NM']           25L      -118.404501         33.935085   
93     RWY 23R     ['3 NM']           23R       -78.792358         35.882177   
94      RWY 10     ['1 NM']            10       -81.206715         32.128607   

    DECLINATION  RWY_BEARING  
0      -4.13012    315.86988  
1       3.06900    183.06900  
2      -1.62738    358.37262  
3      15.40471    175.40471  
4       2.94849    152.94849  
..          ...          ...  
90     -4.96630    265.03370  
91     11.71212    261.71212  
92     11.71212    261.71212  
93     -9.22834    220.77166  
94     -7.20523     92.79477  

[95 rows x 10 columns]

### Convert Distances from NM to Kilometers

In [17]:
# Convert all distance from NM to kilometers
dist_kilo = []

for i in range(len(rwy95_decDF)):
    distanceKilo = int(re.sub("\D","",rwy95_decDF['UAS_Distance'][i]))* 1.852 # converting NM to kilometers
    dist_kilo.append(distanceKilo)
    
rwy95_decDF['Distance_Kilometers'] = dist_kilo

In [18]:
rwy95_decDF

REMARKS UASLOCATION IDENT  \
0   Aircraft observed a black UAS at 1,000 feet wh...         PAM   PAM   
1   Aircraft observed a UAS above while W bound at...         DFW   DFW   
2   Aircraft observed a UAS at the same altitude w...         MEM   MEM   
3   Aircraft observed a red UAS off the right side...         SEA   SEA   
4   Aircraft observed a small red UAS E bound off ...         ADS   ADS   
..                                                ...         ...   ...   
90  Aircraft reported a UAS passing 100 feet below...         TLH   TLH   
91  Aircraft reported a UAS off the right side, ap...         LAX   LAX   
92  Aircraft reported a grey, basketball size, UAS...         LAX   LAX   
93  Aircraft reported a UAS off the left side whil...         RDU   RDU   
94  Aircraft reported a fixed wing UAS pass within...         SAV   SAV   

   RWYLOCATION UAS_Distance RWYDESIGNATOR  Runway.Latitude  Runway.Longitude  \
0      RWY 32R     ['3 NM']           32R       -85.570825         30.068523   
1      RWY 18R     ['5 NM']           18R       -97.054815         32.893723   
2      RWY 36L     ['6 NM']           36L       -89.987278         35.039251   
3      RWY 16R     ['4 NM']           16R      -122.317932         47.449860   
4       RWY 15     ['3 NM']            15       -96.837304         32.970408   
..         ...          ...           ...              ...               ...   
90      RWY 27     ['2 NM']            27       -84.346397         30.391355   
91     RWY 25L     ['3 NM']           25L      -118.404501         33.935085   
92     RWY 25L    ['10 NM']           25L      -118.404501         33.935085   
93     RWY 23R     ['3 NM']           23R       -78.792358         35.882177   
94      RWY 10     ['1 NM']            10       -81.206715         32.128607   

    DECLINATION  RWY_BEARING  Distance_Kilometers  
0      -4.13012    315.86988                5.556  
1       3.06900    183.06900                9.260  
2      -1.62738    358.37262               11.112  
3      15.40471    175.40471                7.408  
4       2.94849    152.94849                5.556  
..          ...          ...                  ...  
90     -4.96630    265.03370                3.704  
91     11.71212    261.71212                5.556  
92     11.71212    261.71212               18.520  
93     -9.22834    220.77166                5.556  
94     -7.20523     92.79477                1.852  

[95 rows x 11 columns]

# SWITCH Runway.Latitude and Runway.Longitude (I THINK ERIC HAS THESE MIXED UP ACCIDENTALLY)

In [19]:
rwy95_decDF.rename(columns = {'Runway.Latitude': 'Runway_Longitude', 'Runway.Longitude' : 'Runway_Latitude'}, inplace = True)

### Calculate UAS Latitude/Longitude Information using geopy library

In [20]:
uas_lat = []
uas_long = []


for i in range(len(rwy95_decDF)):
    lat_rwy = pd.to_numeric(rwy95_decDF['Runway_Latitude'][i])
    long_rwy = pd.to_numeric(rwy95_decDF['Runway_Longitude'][i])
    b = pd.to_numeric(rwy95_decDF['RWY_BEARING'][i])
    d = pd.to_numeric(rwy95_decDF['Distance_Kilometers'][i])

    origin = geopy.Point(lat_rwy, long_rwy)
    destination = geodesic(kilometers=d).destination(origin,b)

    lat2, lon2, = destination.latitude, destination.longitude

    uas_lat.append(lat2)
    uas_long.append(lon2)

# Append UAS Lat/Long information to DataFrame
rwy95_decDF['UAS_Latitude'] = uas_lat
rwy95_decDF['UAS_Longitude'] = uas_long

In [21]:
rwy95_decDF

REMARKS UASLOCATION IDENT  \
0   Aircraft observed a black UAS at 1,000 feet wh...         PAM   PAM   
1   Aircraft observed a UAS above while W bound at...         DFW   DFW   
2   Aircraft observed a UAS at the same altitude w...         MEM   MEM   
3   Aircraft observed a red UAS off the right side...         SEA   SEA   
4   Aircraft observed a small red UAS E bound off ...         ADS   ADS   
..                                                ...         ...   ...   
90  Aircraft reported a UAS passing 100 feet below...         TLH   TLH   
91  Aircraft reported a UAS off the right side, ap...         LAX   LAX   
92  Aircraft reported a grey, basketball size, UAS...         LAX   LAX   
93  Aircraft reported a UAS off the left side whil...         RDU   RDU   
94  Aircraft reported a fixed wing UAS pass within...         SAV   SAV   

   RWYLOCATION UAS_Distance RWYDESIGNATOR  Runway_Longitude  Runway_Latitude  \
0      RWY 32R     ['3 NM']           32R        -85.570825        30.068523   
1      RWY 18R     ['5 NM']           18R        -97.054815        32.893723   
2      RWY 36L     ['6 NM']           36L        -89.987278        35.039251   
3      RWY 16R     ['4 NM']           16R       -122.317932        47.449860   
4       RWY 15     ['3 NM']            15        -96.837304        32.970408   
..         ...          ...           ...               ...              ...   
90      RWY 27     ['2 NM']            27        -84.346397        30.391355   
91     RWY 25L     ['3 NM']           25L       -118.404501        33.935085   
92     RWY 25L    ['10 NM']           25L       -118.404501        33.935085   
93     RWY 23R     ['3 NM']           23R        -78.792358        35.882177   
94      RWY 10     ['1 NM']            10        -81.206715        32.128607   

    DECLINATION  RWY_BEARING  Distance_Kilometers  UAS_Latitude  UAS_Longitude  
0      -4.13012    315.86988                5.556     30.104491     -85.610962  
1       3.06900    183.06900                9.260     32.810345     -97.060108  
2      -1.62738    358.37262               11.112     35.139371     -89.990741  
3      15.40471    175.40471                7.408     47.383443    -122.310072  
4       2.94849    152.94849                5.556     32.925788     -96.810288  
..          ...          ...                  ...           ...            ...  
90     -4.96630    265.03370                3.704     30.388457     -84.384793  
91     11.71212    261.71212                5.556     33.927851    -118.463962  
92     11.71212    261.71212               18.520     33.910858    -118.602666  
93     -9.22834    220.77166                5.556     35.844248     -78.832521  
94     -7.20523     92.79477                1.852     32.127791     -81.187112  

[95 rows x 13 columns]

In [22]:
# Export final new_df to .csv 
rwy95_decDF.to_csv('other2142_runways95_UAS_latlong.csv', index = False)

***
# Exploring the 504 Remaining Remarks Referencing Runways (RWYS)
***
`These Remarks are a subset of the 2142 records that were NOT hit on by the initial regex: 'XX NM dir XXX'.`

In [23]:
rwy_unid504 = pd.DataFrame(rwy599, index = set(rwy599.index).difference(set(rwy95_decDF.index))).reset_index(drop=True)
rwy_unid504

index                                            REMARKS  \
0       95  Aircraft observed a white UAS at 500 feet whil...   
1       96  Aircraft observed a UAS off the right side whi...   
2       97  Aircraft observed a flight of 2 white UAS on t...   
3       98  Aircraft observed a black Quad-copter UAS off ...   
4       99  Aircraft observed a white UAS off the left sid...   
..     ...                                                ...   
499    596  Aircraft reported a fixed wing UAS pass within...   
500    597  Aircraft reported a grey quad-copter UAS off t...   
501    598  Aircraft reported a black quad-copter UAS off ...   
502    599  Aircraft reported a NMAC with a UAS while on a...   
503    600  VFR Aircraft reported a near mid-air with a UA...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0             ['RWY 9L']         NaN   NaN         NaN           []  
1            ['RWY 24R']         NaN   NaN         NaN      ['5NM']  
2             ['RWY 24']         CRQ   CRQ      RWY 24           []  
3             ['RWY17L']         NaN   NaN         NaN           []  
4             ['RWY 8L']         NaN   NaN         NaN      ['3NM']  
..                   ...         ...   ...         ...          ...  
499           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']  
500          ['RWY 27L']         SFB   SFB     RWY 27L           []  
501           ['RWY 3R']         NaN   NaN         NaN           []  
502            ['RWY05']         NaN   NaN         NaN     ['6 NM']  
503          ['RWY 28R']         BOI   BOI     RWY 28R           []  

[504 rows x 7 columns]

# Look at the records for which there is a valid IDENT

`There are 202 records (within the 504 unidentified runway remarks) that have valid IDENT/REPORTINGFACILITY Information associated with the Runway mentioned within the Remark.`

In [24]:
rwy202 = rwy_unid504[rwy_unid504['IDENT'].notnull()].reset_index(drop=True)
rwy202

index                                            REMARKS  \
0       97  Aircraft observed a flight of 2 white UAS on t...   
1      104  Aircraft observed a UAS off the left side whil...   
2      107  Aircraft observed a UAS while NE bound at 900 ...   
3      109  Aircraft observed a UAS at the same altitude w...   
4      111  Aircraft observed a UAS while S bound at 1,000...   
..     ...                                                ...   
197    593  Aircraft reported a UAS at 4,400 feet while on...   
198    594  Aircraft reported a UAS off the left side whil...   
199    596  Aircraft reported a fixed wing UAS pass within...   
200    597  Aircraft reported a grey quad-copter UAS off t...   
201    600  VFR Aircraft reported a near mid-air with a UA...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0             ['RWY 24']         CRQ   CRQ      RWY 24           []  
1            ['RWY 23L']         RDU   RDU     RWY 23L           []  
2             ['RWY 28']         MYF   MYF      RWY 28           []  
3            ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']  
4            ['RWY 26R']         LGB   LGB     RWY 26R           []  
..                   ...         ...   ...         ...          ...  
197          ['RWY 35C']         DFW   DFW     RWY 35C           []  
198          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']  
199           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']  
200          ['RWY 27L']         SFB   SFB     RWY 27L           []  
201          ['RWY 28R']         BOI   BOI     RWY 28R           []  

[202 rows x 7 columns]

In [25]:
# First we need to replace all [] in UAS_Distance field with NaN
rwy202_df = rwy202.mask(rwy202.applymap(str).eq('[]'))
#rwy202_df

## Explore the records for which there is valid UAS_Distance (in NM) information associated with the RWY Location

`There are 100 records (of the 202 with valid Reporting Facility / IDENT information) that also contain a valid UAS Distance (in NM).`

In [26]:
rwy202_dist = rwy202_df[rwy202_df['UAS_Distance'].notnull()].reset_index(drop=True)
rwy202_dist

index                                            REMARKS  \
0     109  Aircraft observed a UAS at the same altitude w...   
1     124  Aircraft observed a red UAS off the right side...   
2     145  Aircraft observed a small red UAS E bound off ...   
3     148  Aircraft observed a UAS, operating at approxim...   
4     160  Aircraft observed a UAS 500 feet below the air...   
..    ...                                                ...   
95    516  Aircraft reported a UAS passing 100 feet below...   
96    588  Aircraft reported a UAS off the right side, ap...   
97    591  Aircraft reported a grey, basketball size, UAS...   
98    594  Aircraft reported a UAS off the left side whil...   
99    596  Aircraft reported a fixed wing UAS pass within...   

   UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0           ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']  
1           ['RWY 16R']         SEA   SEA     RWY 16R     ['4 NM']  
2            ['RWY 15']         ADS   ADS      RWY 15     ['3 NM']  
3           ['RWY 10L']         PDX   PDX     RWY 10L     ['7 NM']  
4           ['RWY 25L']         LAX   LAX     RWY 25L     ['2 NM']  
..                  ...         ...   ...         ...          ...  
95           ['RWY 27']         TLH   TLH      RWY 27     ['2 NM']  
96          ['RWY 25L']         LAX   LAX     RWY 25L     ['3 NM']  
97          ['RWY 25L']         LAX   LAX     RWY 25L    ['10 NM']  
98          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']  
99           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']  

[100 rows x 7 columns]

In [27]:
# Export to .csv 
rwy202_dist.to_csv('rwy100.csv', index = False)

#### Merge rwy202_dist and airports_runways_linked.csv (air_rwy)

`Merge the two datasets on the 'IDENT' field.`

In [28]:
rwy100_airwy = pd.merge(rwy202_dist, air_rwy, on = 'IDENT', how = 'left').reset_index(drop=True)
rwy100_airwy.rename(columns = {'properties.DESIGNATOR': 'DESIGNATOR'}, inplace = True)
rwy100_airwy

index                                            REMARKS  \
0      109  Aircraft observed a UAS at the same altitude w...   
1      109  Aircraft observed a UAS at the same altitude w...   
2      109  Aircraft observed a UAS at the same altitude w...   
3      109  Aircraft observed a UAS at the same altitude w...   
4      124  Aircraft observed a red UAS off the right side...   
..     ...                                                ...   
315    594  Aircraft reported a UAS off the left side whil...   
316    594  Aircraft reported a UAS off the left side whil...   
317    594  Aircraft reported a UAS off the left side whil...   
318    596  Aircraft reported a fixed wing UAS pass within...   
319    596  Aircraft reported a fixed wing UAS pass within...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  \
0            ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']   
1            ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']   
2            ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']   
3            ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']   
4            ['RWY 16R']         SEA   SEA     RWY 16R     ['4 NM']   
..                   ...         ...   ...         ...          ...   
315          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
316          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
317          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
318           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']   
319           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']   

                     properties.GLOBAL_ID  \
0    3FD784E7-DFCC-4115-842D-60DB410A9D25   
1    462D6D8A-8D57-443B-B452-E01E8E40D249   
2    DE9A1353-6BD4-410A-A407-B2A253445358   
3    DC9631DF-1821-449A-AD41-8EA0B8B40375   
4    7AC9CEA7-9C80-44BD-84DC-63EDFA5C4EC2   
..                                    ...   
315  91541445-E5CF-40C1-8E8F-3E759B9CA482   
316  D1A26CC8-C4DA-48CF-A0B0-7FD883EBA2FA   
317  B270C60D-5C0C-41EA-8410-5DA7D520B4E3   
318  B2D0E0C7-E412-458E-B683-52C4DD4E4EA9   
319  FFE4D41C-A5CA-4072-BCE0-59A70192F6D3   

                    properties.AIRPORT_ID DESIGNATOR  Runway.Latitude  \
0    DA96A371-EB42-4312-8EAF-FA237DE1E404      09/27       -89.973792   
1    DA96A371-EB42-4312-8EAF-FA237DE1E404    18C/36C       -89.975969   
2    DA96A371-EB42-4312-8EAF-FA237DE1E404    18L/36R       -89.972799   
3    DA96A371-EB42-4312-8EAF-FA237DE1E404    18R/36L       -89.987278   
4    5338FBAA-B183-4BC0-B3F1-27ABF6D1AD7C    16C/34C      -122.311074   
..                                    ...        ...              ...   
315  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4    05L/23R       -78.792358   
316  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4    05R/23L       -78.790127   
317  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4      14/32       -78.777558   
318  D339C530-1D36-4587-8130-9E9938BF474D      01/19       -81.200005   
319  D339C530-1D36-4587-8130-9E9938BF474D      10/28       -81.206715   

     Runway.Longitude                       NAME   LATITUDE   LONGITUDE  \
0           35.058252               Memphis Intl  35.042411  -89.976679   
1           35.042383               Memphis Intl  35.042411  -89.976679   
2           35.038933               Memphis Intl  35.042411  -89.976679   
3           35.039251               Memphis Intl  35.042411  -89.976679   
4           47.448312        Seattle-Tacoma Intl  47.449889 -122.311778   
..                ...                        ...        ...         ...   
315         35.882177        Raleigh-Durham Intl  35.877639  -78.787472   
316         35.870421        Raleigh-Durham Intl  35.877639  -78.787472   
317         35.870885        Raleigh-Durham Intl  35.877639  -78.787472   
318         32.124275  Savannah/Hilton Head Intl  32.127583  -81.202139   
319         32.128607  Savannah/Hilton Head Intl  32.127583  -81.202139   

    ICAO_ID  
0      KMEM  
1      KMEM  
2      KME

#### Extract Runway Designator from RWYLOCATION field

`We want to be able to obtain the correct latitude/longitude coordinates for the appropriate runway mentioned in the remark.`

In [29]:
rwydesignator = []

for i in range(len(rwy100_airwy)):
    rwy_split = rwy100_airwy['RWYLOCATION'][i].split(' ')
    rwydesignator.append(rwy_split[1])

# Add split column with RWY designators to dataframe as new field
rwy100_airwy['RWYDESIGNATOR'] = rwydesignator

#### Check if RWY designator found in REMARK matches one of the RWY designators from the rwy.csv dataset

`If YES: RWYinREMARK field is 'True'`

`If NO: RWYinREMARK field is 'False'`
***
***NOTE:*** Since we linked on the 'IDENT' field, there are repeated remarks as a result of there being multiple runways at
each airport. Thus, we are only interested in the records for which the 'RWYinREMARK' field is 'True'.

In [30]:
rwy100_airwy['RWYinREMARK'] = rwy100_airwy.apply(lambda x: str(x.RWYDESIGNATOR) in str(x.DESIGNATOR), axis=1)

# Only keep records for which the RWY Designator is present in the DESIGNATOR field 
rwy93 = rwy100_airwy[rwy100_airwy['RWYinREMARK'] == True].reset_index(drop=True)

In [31]:
rwy93

index                                            REMARKS  \
0     109  Aircraft observed a UAS at the same altitude w...   
1     124  Aircraft observed a red UAS off the right side...   
2     145  Aircraft observed a small red UAS E bound off ...   
3     148  Aircraft observed a UAS, operating at approxim...   
4     160  Aircraft observed a UAS 500 feet below the air...   
..    ...                                                ...   
88    516  Aircraft reported a UAS passing 100 feet below...   
89    588  Aircraft reported a UAS off the right side, ap...   
90    591  Aircraft reported a grey, basketball size, UAS...   
91    594  Aircraft reported a UAS off the left side whil...   
92    596  Aircraft reported a fixed wing UAS pass within...   

   UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  \
0           ['RWY 36L']         MEM   MEM     RWY 36L     ['6 NM']   
1           ['RWY 16R']         SEA   SEA     RWY 16R     ['4 NM']   
2            ['RWY 15']         ADS   ADS      RWY 15     ['3 NM']   
3           ['RWY 10L']         PDX   PDX     RWY 10L     ['7 NM']   
4           ['RWY 25L']         LAX   LAX     RWY 25L     ['2 NM']   
..                  ...         ...   ...         ...          ...   
88           ['RWY 27']         TLH   TLH      RWY 27     ['2 NM']   
89          ['RWY 25L']         LAX   LAX     RWY 25L     ['3 NM']   
90          ['RWY 25L']         LAX   LAX     RWY 25L    ['10 NM']   
91          ['RWY 23R']         RDU   RDU     RWY 23R     ['3 NM']   
92           ['RWY 10']         SAV   SAV      RWY 10     ['1 NM']   

                    properties.GLOBAL_ID  \
0   DC9631DF-1821-449A-AD41-8EA0B8B40375   
1   D8DB5055-844E-44D9-A4DA-C7E399892BE1   
2   202E6282-A103-44B0-ACCE-3D984B66D091   
3   DCC8037A-6FE3-4A15-8A3B-3E7253D8FE19   
4   8ED22315-4327-432F-9C74-65E1A06D97A5   
..                                   ...   
88  6AFB0072-D2C9-4832-9B3A-9BA008C1E9F9   
89  8ED22315-4327-432F-9C74-65E1A06D97A5   
90  8ED22315-4327-432F-9C74-65E1A06D97A5   
91  91541445-E5CF-40C1-8E8F-3E759B9CA482   
92  FFE4D41C-A5CA-4072-BCE0-59A70192F6D3   

                   properties.AIRPORT_ID DESIGNATOR  Runway.Latitude  \
0   DA96A371-EB42-4312-8EAF-FA237DE1E404    18R/36L       -89.987278   
1   5338FBAA-B183-4BC0-B3F1-27ABF6D1AD7C    16R/34L      -122.317932   
2   3646423B-964B-466A-9CE9-3CB43457F772      15/33       -96.837304   
3   3B44B4F3-705A-4A4D-AEBC-19BAF63DA508    10L/28R      -122.586637   
4   2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
..                                   ...        ...              ...   
88  265A7B00-46FF-4744-8673-499589FFA2BF      09/27       -84.346397   
89  2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
90  2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
91  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4    05L/23R       -78.792358   
92  D339C530-1D36-4587-8130-9E9938BF474D      10/28       -81.206715   

    Runway.Longitude                       NAME   LATITUDE   LONGITUDE  \
0          35.039251               Memphis Intl  35.042411  -89.976679   
1          47.449860        Seattle-Tacoma Intl  47.449889 -122.311778   
2          32.970408                    Addison  32.968556  -96.836444   
3          45.591254              Portland Intl  45.588709 -122.596869   
4          33.935085           Los Angeles Intl  33.942496 -118.408049   
..               ...                        ...        ...         ...   
88         30.391355           Tallahassee Intl  30.396754  -84.350869   
89         33.935085           Los Angeles Intl  33.942496 -118.408049   
90         33.935085           Los Angeles Intl  33.942496 -118.408049   
91         35.882177        Raleigh-Durham Intl  35.877639  -78.787472   
92         32.128607  Savannah/Hilton Head Intl  32.127583  -81.202139   

   ICAO_ID RWYDESIGNATOR  RWYinREMARK  
0     KMEM           36L         True  
1     KSEA           16R       

#### Merge air_dec with rwy93

`Merge the two datasets on the 'IDENT' field.`

`air_dec --> Contains the cleaned airports supplemental dataset with the Declination value at the airport`

`rwy93 --> Dataframe containing 93 records referencing Runways that contain all relevant information (from the 100 records containing UAS_Distance information)`

In [32]:
rwy93_dec = pd.merge(rwy93, air_dec, on = 'IDENT', how = 'left')

# Only keep necessary columns/fields of interest
rwy93_decDF = pd.DataFrame(rwy93_dec[['REMARKS', 'UASLOCATION', 'IDENT', 'RWYLOCATION', 'UAS_Distance', 'RWYDESIGNATOR', 'Runway.Latitude', 'Runway.Longitude', 'DECLINATION']])

### Calculate Bearing Information for Runways

`To do this, you take the number portion of the Runway Designator (in the 'RWYDESIGNATOR' field) and multiply it by 10 (effectively adding a 0 to the end) and add the DECLINATION to this number`

***Example:*** *RWY 32R has designator 32R which indicates a bearing of 320 degrees*

In [33]:
designator = rwy93_decDF['RWYDESIGNATOR']
designator_bearing = []

for i in range(len(rwy93_decDF)):
    designator_bearing.append(int(re.sub("\D","",designator[i]))*10 + rwy93_decDF['DECLINATION'][i])

rwy93_decDF['RWY_BEARING'] = designator_bearing

### Convert Distances from NM to Kilometers

In [34]:
# Convert all distance from NM to kilometers
dist_kilo = []

for i in range(len(rwy93_decDF)):
    distanceKilo = int(re.sub("\D","",rwy93_decDF['UAS_Distance'][i]))* 1.852 # converting NM to kilometers
    dist_kilo.append(distanceKilo)
    
rwy93_decDF['Distance_Kilometers'] = dist_kilo

In [35]:
rwy93_decDF.rename(columns = {'Runway.Latitude': 'Runway_Longitude', 'Runway.Longitude' : 'Runway_Latitude'}, inplace = True)

## Calculate UAS Latitude/Longitude Information using geopy library

In [36]:
uas_lat = []
uas_long = []


for i in range(len(rwy93_decDF)):
    lat_rwy = pd.to_numeric(rwy93_decDF['Runway_Latitude'][i])
    long_rwy = pd.to_numeric(rwy93_decDF['Runway_Longitude'][i])
    b = pd.to_numeric(rwy93_decDF['RWY_BEARING'][i])
    d = pd.to_numeric(rwy93_decDF['Distance_Kilometers'][i])

    origin = geopy.Point(lat_rwy, long_rwy)
    destination = geodesic(kilometers=d).destination(origin,b)

    lat2, lon2, = destination.latitude, destination.longitude

    uas_lat.append(lat2)
    uas_long.append(lon2)

# Append UAS Lat/Long information to DataFrame
rwy93_decDF['UAS_Latitude'] = uas_lat
rwy93_decDF['UAS_Longitude'] = uas_long

In [37]:
rwy93_decDF

REMARKS UASLOCATION IDENT  \
0   Aircraft observed a UAS at the same altitude w...         MEM   MEM   
1   Aircraft observed a red UAS off the right side...         SEA   SEA   
2   Aircraft observed a small red UAS E bound off ...         ADS   ADS   
3   Aircraft observed a UAS, operating at approxim...         PDX   PDX   
4   Aircraft observed a UAS 500 feet below the air...         LAX   LAX   
..                                                ...         ...   ...   
88  Aircraft reported a UAS passing 100 feet below...         TLH   TLH   
89  Aircraft reported a UAS off the right side, ap...         LAX   LAX   
90  Aircraft reported a grey, basketball size, UAS...         LAX   LAX   
91  Aircraft reported a UAS off the left side whil...         RDU   RDU   
92  Aircraft reported a fixed wing UAS pass within...         SAV   SAV   

   RWYLOCATION UAS_Distance RWYDESIGNATOR  Runway_Longitude  Runway_Latitude  \
0      RWY 36L     ['6 NM']           36L        -89.987278        35.039251   
1      RWY 16R     ['4 NM']           16R       -122.317932        47.449860   
2       RWY 15     ['3 NM']            15        -96.837304        32.970408   
3      RWY 10L     ['7 NM']           10L       -122.586637        45.591254   
4      RWY 25L     ['2 NM']           25L       -118.404501        33.935085   
..         ...          ...           ...               ...              ...   
88      RWY 27     ['2 NM']            27        -84.346397        30.391355   
89     RWY 25L     ['3 NM']           25L       -118.404501        33.935085   
90     RWY 25L    ['10 NM']           25L       -118.404501        33.935085   
91     RWY 23R     ['3 NM']           23R        -78.792358        35.882177   
92      RWY 10     ['1 NM']            10        -81.206715        32.128607   

    DECLINATION  RWY_BEARING  Distance_Kilometers  UAS_Latitude  UAS_Longitude  
0      -1.62738    358.37262               11.112     35.139371     -89.990741  
1      15.40471    175.40471                7.408     47.383443    -122.310072  
2       2.94849    152.94849                5.556     32.925788     -96.810288  
3      15.01885    115.01885               12.964     45.541825    -122.436221  
4      11.71212    261.71212                3.704     33.930265    -118.444143  
..          ...          ...                  ...           ...            ...  
88     -4.96630    265.03370                3.704     30.388457     -84.384793  
89     11.71212    261.71212                5.556     33.927851    -118.463962  
90     11.71212    261.71212               18.520     33.910858    -118.602666  
91     -9.22834    220.77166                5.556     35.844248     -78.832521  
92     -7.20523     92.79477                1.852     32.127791     -81.187112  

[93 rows x 13 columns]

In [38]:
# Export to .csv 
rwy93_decDF.to_csv('rwy93_uasLatLong.csv', index = False)

#### Obtain Dataframe (and corresponding .csv file) for the 7 records for which UAS Lat/Long information was not captured

In [39]:
rwy_unid7 = pd.DataFrame(rwy202_dist, index = set(rwy202_dist.index).difference(set(rwy93_decDF.index))).reset_index(drop=True)

In [40]:
rwy_unid7

index                                            REMARKS  \
0    588  Aircraft reported a UAS off the right side, ap...   
1    591  Aircraft reported a grey, basketball size, UAS...   
2    594  Aircraft reported a UAS off the left side whil...   
3    596  Aircraft reported a fixed wing UAS pass within...   
4    467  Aircraft reported a UAS off their left side at...   
5    511  Aircraft reported a red fixed wing UAS, 200 fe...   
6    516  Aircraft reported a UAS passing 100 feet below...   

  UAS_Location_Runways   UASLOCATION IDENT RWYLOCATION UAS_Distance  
0          ['RWY 25L']           LAX   LAX     RWY 25L     ['3 NM']  
1          ['RWY 25L']           LAX   LAX     RWY 25L    ['10 NM']  
2          ['RWY 23R']           RDU   RDU     RWY 23R     ['3 NM']  
3           ['RWY 10']           SAV   SAV      RWY 10     ['1 NM']  
4           ['RWY 31']  2 NM SE  LGA   LGA      RWY 31     ['2 NM']  
5           ['RWY 24']           WOC   WOC      RWY 24     ['4 NM']  
6           ['RWY 27']           TLH   TLH      RWY 27     ['2 NM']

In [41]:
# Export to .csv 
rwy_unid7.to_csv('rwy7_EXCEPTIONS.csv', index = False)

***

## Explore the records for which there is NO valid UAS_Distance (in NM) information associated with the RWY Location

`There are 102 records (of the 202 with valid Reporting Facility / IDENT information) that DO NOT contain a valid UAS Distance (in NM).`

In [42]:
rwy202_noDist = rwy202_df[rwy202_df['UAS_Distance'].isnull()].reset_index(drop=True)
rwy202_noDist

index                                            REMARKS  \
0       97  Aircraft observed a flight of 2 white UAS on t...   
1      104  Aircraft observed a UAS off the left side whil...   
2      107  Aircraft observed a UAS while NE bound at 900 ...   
3      111  Aircraft observed a UAS while S bound at 1,000...   
4      112  Aircraft observed a UAS at the 2 o'clock posit...   
..     ...                                                ...   
97     589  Aircraft reported a red and white quad-copter ...   
98     590  Aircraft reported a UAS off left side, while o...   
99     593  Aircraft reported a UAS at 4,400 feet while on...   
100    597  Aircraft reported a grey quad-copter UAS off t...   
101    600  VFR Aircraft reported a near mid-air with a UA...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0             ['RWY 24']         CRQ   CRQ      RWY 24          NaN  
1            ['RWY 23L']         RDU   RDU     RWY 23L          NaN  
2             ['RWY 28']         MYF   MYF      RWY 28          NaN  
3            ['RWY 26R']         LGB   LGB     RWY 26R          NaN  
4            ['RWY 19R']         MKE   MKE     RWY 19R          NaN  
..                   ...         ...   ...         ...          ...  
97            ['RWY 15']         PNE   PNE      RWY 15          NaN  
98           ['RWY 35L']         DWH   DWH     RWY 35L          NaN  
99           ['RWY 35C']         DFW   DFW     RWY 35C          NaN  
100          ['RWY 27L']         SFB   SFB     RWY 27L          NaN  
101          ['RWY 28R']         BOI   BOI     RWY 28R          NaN  

[102 rows x 7 columns]

In [43]:
# Export to .csv 
rwy202_noDist.to_csv('rwy102.csv', index = False)

#### Check for Remarks that have substring 'XX mile final' before mention of a Runway

`Be sure to account for multiple types of XX : e.g. 1.5, 05, 14, 0.75, etc.`

In [44]:
# Create List that contains each Standard Remark, and the Heading/Direction
# information contained in each remark

remark_uas_loc = []
remarks = rwy202_noDist['REMARKS']

# regular expression for any heading/direction
headir_regex = '\.?[0-9]\.?[0-9]*[0-9]*\s?mile?[s]?\s?final?'

# Loop through all remarks and search for the heading/direction regex above
for i in range(len(remarks)):
    head_dir = re.findall(headir_regex, remarks[i])
    remark_uas_loc.append(remarks[i])
    remark_uas_loc.append(head_dir)

# Split Remarks and Heading/Directions into two seperate lists and create
# pandas dataframe
remark = []
uas_loc = []

for i in range(0, len(remark_uas_loc), 2):
    remark.append(remark_uas_loc[i])
    uas_loc.append(remark_uas_loc[i+1])

rwy202_noDist['UAS_Distance'] = uas_loc

In [45]:
rwy202_noDist

index                                            REMARKS  \
0       97  Aircraft observed a flight of 2 white UAS on t...   
1      104  Aircraft observed a UAS off the left side whil...   
2      107  Aircraft observed a UAS while NE bound at 900 ...   
3      111  Aircraft observed a UAS while S bound at 1,000...   
4      112  Aircraft observed a UAS at the 2 o'clock posit...   
..     ...                                                ...   
97     589  Aircraft reported a red and white quad-copter ...   
98     590  Aircraft reported a UAS off left side, while o...   
99     593  Aircraft reported a UAS at 4,400 feet while on...   
100    597  Aircraft reported a grey quad-copter UAS off t...   
101    600  VFR Aircraft reported a near mid-air with a UA...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION      UAS_Distance  
0             ['RWY 24']         CRQ   CRQ      RWY 24                []  
1            ['RWY 23L']         RDU   RDU     RWY 23L                []  
2             ['RWY 28']         MYF   MYF      RWY 28                []  
3            ['RWY 26R']         LGB   LGB     RWY 26R                []  
4            ['RWY 19R']         MKE   MKE     RWY 19R  [1.5 mile final]  
..                   ...         ...   ...         ...               ...  
97            ['RWY 15']         PNE   PNE      RWY 15                []  
98           ['RWY 35L']         DWH   DWH     RWY 35L                []  
99           ['RWY 35C']         DFW   DFW     RWY 35C                []  
100          ['RWY 27L']         SFB   SFB     RWY 27L                []  
101          ['RWY 28R']         BOI   BOI     RWY 28R                []  

[102 rows x 7 columns]

`There are 12 records (of the 102 with No Distance information) that actually contain distance information, but in the format:
'XX mile final'`

In [46]:
# First we need to replace all [] in UAS_Distance field with NaN
rwy102_df = rwy202_noDist.mask(rwy202_noDist.applymap(str).eq('[]'))
rwy102_df_MF = rwy102_df[rwy102_df['UAS_Distance'].notnull()].reset_index(drop=True)
rwy102_df_MF

index                                            REMARKS  \
0     112  Aircraft observed a UAS at the 2 o'clock posit...   
1     132  Aircraft observed a  silver quadcopter UAS fro...   
2     157  Aircraft observed a UAS SW bound at 900 feet o...   
3     162  Aircraft observed a silver UAS on 4 mile final...   
4     337  Aircraft observed a UAS while descending throu...   
5     338  Aircraft observed a UAS while descending throu...   
6     359  Aircraft observed a UAS 800 feet overhead off ...   
7     364  Aircraft observed a black and silver UAS from ...   
8     366  Aircraft observed a UAS off the left at 2,400 ...   
9     376  Aircraft observed a white UAS off the left sid...   
10    392  Aircraft observed a UAS from the 12 o'clock wh...   
11    563  Aircraft reported a UAS at 1500 ft. while on 3...   

   UAS_Location_Runways UASLOCATION IDENT RWYLOCATION      UAS_Distance  
0           ['RWY 19R']         MKE   MKE     RWY 19R  [1.5 mile final]  
1           ['RWY 17R']         AUS   AUS     RWY 17R    [7 mile final]  
2            ['RWY 22']         LGA   LGA      RWY 22  [2.5 mile final]  
3           ['RWY 31R']         JFK   JFK     RWY 31R    [4 mile final]  
4            ['RWY 24']         ACK   ACK         NaN    [2 mile final]  
5            ['RWY 24']         ACK   ACK         NaN    [2 mile final]  
6            ['RWY 13']         SGJ   SGJ      RWY 13    [1 mile final]  
7           ['RWY 25L']         LAX   LAX     RWY 25L    [2 mile final]  
8           ['RWY 35L']         LAX   LAX     RWY 35L    [7 mile final]  
9           ['RWY 25L']         LAX   LAX     RWY 25L    [3 mile final]  
10          ['RWY 22R']         EWR   EWR     RWY 22R    [3 mile final]  
11          ['RWY 25R']         ENW   ENW     RWY 25R    [3 mile final]

#### Convert Distances from NM to Kilometers

`Using the 'UAS_Distnace' field from rwy102_df_MF`

In [47]:
# Convert all distance from NM to kilometers
dist_kilo = []

for i in range(len(rwy102_df_MF)):
    distanceKilo = float(re.sub(" mile final","",rwy102_df_MF['UAS_Distance'][i][0]))* 1.852 # converting NM to kilometers
    dist_kilo.append(distanceKilo)
    
rwy102_df_MF['Distance_Kilometers'] = dist_kilo
rwy102_df_MF

index                                            REMARKS  \
0     112  Aircraft observed a UAS at the 2 o'clock posit...   
1     132  Aircraft observed a  silver quadcopter UAS fro...   
2     157  Aircraft observed a UAS SW bound at 900 feet o...   
3     162  Aircraft observed a silver UAS on 4 mile final...   
4     337  Aircraft observed a UAS while descending throu...   
5     338  Aircraft observed a UAS while descending throu...   
6     359  Aircraft observed a UAS 800 feet overhead off ...   
7     364  Aircraft observed a black and silver UAS from ...   
8     366  Aircraft observed a UAS off the left at 2,400 ...   
9     376  Aircraft observed a white UAS off the left sid...   
10    392  Aircraft observed a UAS from the 12 o'clock wh...   
11    563  Aircraft reported a UAS at 1500 ft. while on 3...   

   UAS_Location_Runways UASLOCATION IDENT RWYLOCATION      UAS_Distance  \
0           ['RWY 19R']         MKE   MKE     RWY 19R  [1.5 mile final]   
1           ['RWY 17R']         AUS   AUS     RWY 17R    [7 mile final]   
2            ['RWY 22']         LGA   LGA      RWY 22  [2.5 mile final]   
3           ['RWY 31R']         JFK   JFK     RWY 31R    [4 mile final]   
4            ['RWY 24']         ACK   ACK         NaN    [2 mile final]   
5            ['RWY 24']         ACK   ACK         NaN    [2 mile final]   
6            ['RWY 13']         SGJ   SGJ      RWY 13    [1 mile final]   
7           ['RWY 25L']         LAX   LAX     RWY 25L    [2 mile final]   
8           ['RWY 35L']         LAX   LAX     RWY 35L    [7 mile final]   
9           ['RWY 25L']         LAX   LAX     RWY 25L    [3 mile final]   
10          ['RWY 22R']         EWR   EWR     RWY 22R    [3 mile final]   
11          ['RWY 25R']         ENW   ENW     RWY 25R    [3 mile final]   

    Distance_Kilometers  
0                 2.778  
1                12.964  
2                 4.630  
3                 7.408  
4                 3.704  
5                 3.704  
6                 1.852  
7                 3.704  
8                12.964  
9                 5.556  
10                5.556  
11                5.556

In [48]:
# Drop 'RWYLOCATION' field as it has two columns which are 'NaN' even though there is a valid RWY mentioned in the remark
rwy102_df_MF = rwy102_df_MF.drop(columns = ['RWYLOCATION'])
rwy102_df_MF = rwy102_df_MF.drop_duplicates(subset = 'REMARKS')
rwy102_df_MF

index                                            REMARKS  \
0     112  Aircraft observed a UAS at the 2 o'clock posit...   
1     132  Aircraft observed a  silver quadcopter UAS fro...   
2     157  Aircraft observed a UAS SW bound at 900 feet o...   
3     162  Aircraft observed a silver UAS on 4 mile final...   
4     337  Aircraft observed a UAS while descending throu...   
5     338  Aircraft observed a UAS while descending throu...   
6     359  Aircraft observed a UAS 800 feet overhead off ...   
7     364  Aircraft observed a black and silver UAS from ...   
8     366  Aircraft observed a UAS off the left at 2,400 ...   
9     376  Aircraft observed a white UAS off the left sid...   
10    392  Aircraft observed a UAS from the 12 o'clock wh...   
11    563  Aircraft reported a UAS at 1500 ft. while on 3...   

   UAS_Location_Runways UASLOCATION IDENT      UAS_Distance  \
0           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
1           ['RWY 17R']         AUS   AUS    [7 mile final]   
2            ['RWY 22']         LGA   LGA  [2.5 mile final]   
3           ['RWY 31R']         JFK   JFK    [4 mile final]   
4            ['RWY 24']         ACK   ACK    [2 mile final]   
5            ['RWY 24']         ACK   ACK    [2 mile final]   
6            ['RWY 13']         SGJ   SGJ    [1 mile final]   
7           ['RWY 25L']         LAX   LAX    [2 mile final]   
8           ['RWY 35L']         LAX   LAX    [7 mile final]   
9           ['RWY 25L']         LAX   LAX    [3 mile final]   
10          ['RWY 22R']         EWR   EWR    [3 mile final]   
11          ['RWY 25R']         ENW   ENW    [3 mile final]   

    Distance_Kilometers  
0                 2.778  
1                12.964  
2                 4.630  
3                 7.408  
4                 3.704  
5                 3.704  
6                 1.852  
7                 3.704  
8                12.964  
9                 5.556  
10                5.556  
11                5.556

#### Merge rwy102_df_MF and airports_runways_linked.csv (air_rwy)

`Merge the two datasets on the 'IDENT' field.`

In [49]:
rwy102_airwy = pd.merge(rwy102_df_MF, air_rwy, on = 'IDENT', how = 'left').reset_index(drop=True)
rwy102_airwy.rename(columns = {'properties.DESIGNATOR': 'DESIGNATOR'}, inplace = True)
rwy102_airwy

index                                            REMARKS  \
0     112  Aircraft observed a UAS at the 2 o'clock posit...   
1     112  Aircraft observed a UAS at the 2 o'clock posit...   
2     112  Aircraft observed a UAS at the 2 o'clock posit...   
3     112  Aircraft observed a UAS at the 2 o'clock posit...   
4     112  Aircraft observed a UAS at the 2 o'clock posit...   
5     132  Aircraft observed a  silver quadcopter UAS fro...   
6     132  Aircraft observed a  silver quadcopter UAS fro...   
7     132  Aircraft observed a  silver quadcopter UAS fro...   
8     132  Aircraft observed a  silver quadcopter UAS fro...   
9     132  Aircraft observed a  silver quadcopter UAS fro...   
10    157  Aircraft observed a UAS SW bound at 900 feet o...   
11    157  Aircraft observed a UAS SW bound at 900 feet o...   
12    157  Aircraft observed a UAS SW bound at 900 feet o...   
13    157  Aircraft observed a UAS SW bound at 900 feet o...   
14    162  Aircraft observed a silver UAS on 4 mile final...   
15    162  Aircraft observed a silver UAS on 4 mile final...   
16    162  Aircraft observed a silver UAS on 4 mile final...   
17    162  Aircraft observed a silver UAS on 4 mile final...   
18    337  Aircraft observed a UAS while descending throu...   
19    337  Aircraft observed a UAS while descending throu...   
20    337  Aircraft observed a UAS while descending throu...   
21    338  Aircraft observed a UAS while descending throu...   
22    338  Aircraft observed a UAS while descending throu...   
23    338  Aircraft observed a UAS while descending throu...   
24    359  Aircraft observed a UAS 800 feet overhead off ...   
25    359  Aircraft observed a UAS 800 feet overhead off ...   
26    359  Aircraft observed a UAS 800 feet overhead off ...   
27    359  Aircraft observed a UAS 800 feet overhead off ...   
28    359  Aircraft observed a UAS 800 feet overhead off ...   
29    359  Aircraft observed a UAS 800 feet overhead off ...   
30    364  Aircraft observed a black and silver UAS from ...   
31    364  Aircraft observed a black and silver UAS from ...   
32    364  Aircraft observed a black and silver UAS from ...   
33    364  Aircraft observed a black and silver UAS from ...   
34    366  Aircraft observed a UAS off the left at 2,400 ...   
35    366  Aircraft observed a UAS off the left at 2,400 ...   
36    366  Aircraft observed a UAS off the left at 2,400 ...   
37    366  Aircraft observed a UAS off the left at 2,400 ...   
38    376  Aircraft observed a white UAS off the left sid...   
39    376  Aircraft observed a white UAS off the left sid...   
40    376  Aircraft observed a white UAS off the left sid...   
41    376  Aircraft observed a white UAS off the left sid...   
42    392  Aircraft observed a UAS from the 12 o'clock wh...   
43    392  Aircraft observed a UAS from the 12 o'clock wh...   
44    392  Aircraft observed a UAS from the 12 o'clock wh...   
45    392  Aircraft observed a UAS from the 12 o'clock wh...   
46    563  Aircraft reported a UAS at 1500 ft. while on 3...   
47    563  Aircraft reported a UAS at 1500 ft. while on 3...   
48    563  Aircraft reported a UAS at 1500 ft. while on 3...   

   UAS_Location_Runways UASLOCATION IDENT      UAS_Distance  \
0           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
1           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
2           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
3           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
4           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
5           ['RWY 17R']         AUS   AUS    [7 mile final]   
6           ['RWY 17R']         AUS   AUS    [7 mile final]   
7           ['RWY 17R']         AUS   AUS    [7 mile final]   
8           ['RWY 17R']         AUS   AUS    [7 mile final]   
9           ['RWY 17R']         AUS   AUS    [7 mile final]   
10           ['RWY 22']         LGA   LGA  [2.5 mile final]   
11           ['RWY 22']         LGA   LGA  [2.5

#### Extract Runway Designator from UAS_Location_Runways field

`We want to be able to obtain the correct latitude/longitude coordinates for the appropriate runway mentioned in the remark.`

In [50]:
rwydesignator = []

for i in range(len(rwy102_airwy)):
    rwy_split = rwy102_airwy['UAS_Location_Runways'][i].split(' ')
    rwydesignator.append(rwy_split[1][:-2])

# Add split column with RWY designators to dataframe as new field
rwy102_airwy['RWYDESIGNATOR'] = rwydesignator

In [51]:
rwy102_airwy

index                                            REMARKS  \
0     112  Aircraft observed a UAS at the 2 o'clock posit...   
1     112  Aircraft observed a UAS at the 2 o'clock posit...   
2     112  Aircraft observed a UAS at the 2 o'clock posit...   
3     112  Aircraft observed a UAS at the 2 o'clock posit...   
4     112  Aircraft observed a UAS at the 2 o'clock posit...   
5     132  Aircraft observed a  silver quadcopter UAS fro...   
6     132  Aircraft observed a  silver quadcopter UAS fro...   
7     132  Aircraft observed a  silver quadcopter UAS fro...   
8     132  Aircraft observed a  silver quadcopter UAS fro...   
9     132  Aircraft observed a  silver quadcopter UAS fro...   
10    157  Aircraft observed a UAS SW bound at 900 feet o...   
11    157  Aircraft observed a UAS SW bound at 900 feet o...   
12    157  Aircraft observed a UAS SW bound at 900 feet o...   
13    157  Aircraft observed a UAS SW bound at 900 feet o...   
14    162  Aircraft observed a silver UAS on 4 mile final...   
15    162  Aircraft observed a silver UAS on 4 mile final...   
16    162  Aircraft observed a silver UAS on 4 mile final...   
17    162  Aircraft observed a silver UAS on 4 mile final...   
18    337  Aircraft observed a UAS while descending throu...   
19    337  Aircraft observed a UAS while descending throu...   
20    337  Aircraft observed a UAS while descending throu...   
21    338  Aircraft observed a UAS while descending throu...   
22    338  Aircraft observed a UAS while descending throu...   
23    338  Aircraft observed a UAS while descending throu...   
24    359  Aircraft observed a UAS 800 feet overhead off ...   
25    359  Aircraft observed a UAS 800 feet overhead off ...   
26    359  Aircraft observed a UAS 800 feet overhead off ...   
27    359  Aircraft observed a UAS 800 feet overhead off ...   
28    359  Aircraft observed a UAS 800 feet overhead off ...   
29    359  Aircraft observed a UAS 800 feet overhead off ...   
30    364  Aircraft observed a black and silver UAS from ...   
31    364  Aircraft observed a black and silver UAS from ...   
32    364  Aircraft observed a black and silver UAS from ...   
33    364  Aircraft observed a black and silver UAS from ...   
34    366  Aircraft observed a UAS off the left at 2,400 ...   
35    366  Aircraft observed a UAS off the left at 2,400 ...   
36    366  Aircraft observed a UAS off the left at 2,400 ...   
37    366  Aircraft observed a UAS off the left at 2,400 ...   
38    376  Aircraft observed a white UAS off the left sid...   
39    376  Aircraft observed a white UAS off the left sid...   
40    376  Aircraft observed a white UAS off the left sid...   
41    376  Aircraft observed a white UAS off the left sid...   
42    392  Aircraft observed a UAS from the 12 o'clock wh...   
43    392  Aircraft observed a UAS from the 12 o'clock wh...   
44    392  Aircraft observed a UAS from the 12 o'clock wh...   
45    392  Aircraft observed a UAS from the 12 o'clock wh...   
46    563  Aircraft reported a UAS at 1500 ft. while on 3...   
47    563  Aircraft reported a UAS at 1500 ft. while on 3...   
48    563  Aircraft reported a UAS at 1500 ft. while on 3...   

   UAS_Location_Runways UASLOCATION IDENT      UAS_Distance  \
0           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
1           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
2           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
3           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
4           ['RWY 19R']         MKE   MKE  [1.5 mile final]   
5           ['RWY 17R']         AUS   AUS    [7 mile final]   
6           ['RWY 17R']         AUS   AUS    [7 mile final]   
7           ['RWY 17R']         AUS   AUS    [7 mile final]   
8           ['RWY 17R']         AUS   AUS    [7 mile final]   
9           ['RWY 17R']         AUS   AUS    [7 mile final]   
10           ['RWY 22']         LGA   LGA  [2.5 mile final]   
11           ['RWY 22']         LGA   LGA  [2.5

#### Check if RWY designator found in REMARK matches one of the RWY designators from the rwy.csv dataset

`If YES: RWYinREMARK field is 'True'`

`If NO: RWYinREMARK field is 'False'`
***
***NOTE:*** Since we linked on the 'IDENT' field, there are repeated remarks as a result of there being multiple runways at
each airport. Thus, we are only interested in the records for which the 'RWYinREMARK' field is 'True'.

In [52]:
rwy102_airwy['RWYinREMARK'] = rwy102_airwy.apply(lambda x: str(x.RWYDESIGNATOR) in str(x.DESIGNATOR), axis=1)

# Only keep records for which the RWY Designator is present in the DESIGNATOR field 
rwy10 = rwy102_airwy[rwy102_airwy['RWYinREMARK'] == True].drop_duplicates(subset = ['REMARKS']).reset_index(drop=True)

In [53]:
rwy10

index                                            REMARKS  \
0    112  Aircraft observed a UAS at the 2 o'clock posit...   
1    157  Aircraft observed a UAS SW bound at 900 feet o...   
2    162  Aircraft observed a silver UAS on 4 mile final...   
3    337  Aircraft observed a UAS while descending throu...   
4    338  Aircraft observed a UAS while descending throu...   
5    359  Aircraft observed a UAS 800 feet overhead off ...   
6    364  Aircraft observed a black and silver UAS from ...   
7    376  Aircraft observed a white UAS off the left sid...   
8    392  Aircraft observed a UAS from the 12 o'clock wh...   
9    563  Aircraft reported a UAS at 1500 ft. while on 3...   

  UAS_Location_Runways UASLOCATION IDENT      UAS_Distance  \
0          ['RWY 19R']         MKE   MKE  [1.5 mile final]   
1           ['RWY 22']         LGA   LGA  [2.5 mile final]   
2          ['RWY 31R']         JFK   JFK    [4 mile final]   
3           ['RWY 24']         ACK   ACK    [2 mile final]   
4           ['RWY 24']         ACK   ACK    [2 mile final]   
5           ['RWY 13']         SGJ   SGJ    [1 mile final]   
6          ['RWY 25L']         LAX   LAX    [2 mile final]   
7          ['RWY 25L']         LAX   LAX    [3 mile final]   
8          ['RWY 22R']         EWR   EWR    [3 mile final]   
9          ['RWY 25R']         ENW   ENW    [3 mile final]   

   Distance_Kilometers                  properties.GLOBAL_ID  \
0                2.778  5DFFD5A2-9325-4D97-8BC6-05BC6E6D5B9C   
1                4.630  E79935ED-5475-48DD-85FD-06D6D84D6AB4   
2                7.408  A2421FD0-7187-4324-A544-14659D6BCCBE   
3                3.704  C7EC798C-CBD0-4528-99CA-252048E8B927   
4                3.704  C7EC798C-CBD0-4528-99CA-252048E8B927   
5                1.852  F5F77FD6-9379-4988-B64B-05095A2BCCB0   
6                3.704  8ED22315-4327-432F-9C74-65E1A06D97A5   
7                5.556  8ED22315-4327-432F-9C74-65E1A06D97A5   
8                5.556  EA4D9B89-9E47-4056-B67F-4B17812C60DB   
9                5.556  AE932899-2A43-42D5-8AD6-DDFBF1EBF662   

                  properties.AIRPORT_ID DESIGNATOR  Runway.Latitude  \
0  1050EB1A-2F00-4552-822E-22DAFB9B7F76    01L/19R       -87.895715   
1  E866CAA3-781C-4162-8C31-D16CC9E3E92E      04/22       -73.878696   
2  DD5FB154-0787-4B0F-8554-C033FAFA0155    13L/31R       -73.777891   
3  07ABFEB8-85A2-4FCB-BA5F-583803FE3FEB      06/24       -70.066003   
4  07ABFEB8-85A2-4FCB-BA5F-583803FE3FEB      06/24       -70.066003   
5  556BB8E8-4A7D-4846-A3A2-632E4F29C832      13/31       -81.338523   
6  2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
7  2C5866F1-5E27-4032-9AF8-988CE5DB82BA    07R/25L      -118.404501   
8  6AB3CFCC-F6EE-4B71-A201-2DCA9BB7F2F1    04L/22R       -74.172492   
9  9F9B4071-CDD9-4647-8EFF-71E8432B7D35    07L/25R       -87.932913   

   Runway.Longitude                    NAME   LATITUDE   LONGITUDE ICAO_ID  \
0         42.941378   General Mitchell Intl  42.946932  -87.897064    KMKE   
1         40.775660               Laguardia  40.777250  -73.872611    KLGA   
2         40.652110     John F Kennedy Intl  40.639928  -73.778693    KJFK   
3         41.251812          Nantucket Meml  41.252975  -70.059911    KACK   
4         41.251812          Nantucket Meml  41.252975  -70.059911    KACK   
5         29.959573  Northeast Florida Rgnl  29.959252  -81.339732    KSGJ   
6         33.935085        Los Angeles Intl  33.942496 -118.408049    KLAX   
7         33.935085        Los Angeles Intl  33.942496 -118.408049    KLAX   
8         40.686242     Newark Liberty Intl  40.692480  -74.168687    KEWR   
9         42.596187            Kenosha Rgnl  42.596063  -87.927324    KENW   

  RWYDESIGNATOR  RWYinREMARK  
0           19R         True  
1            22         True  
2           31R         True  
3            24         True  
4            24         True  
5            13         True  
6           25L         True  
7           25L         True  
8

#### Merge air_dec with rwy10

`Merge the two datasets on the 'IDENT' field.`

`air_dec --> Contains the cleaned airports supplemental dataset with the Declination value at the airport`

`rwy10 --> Dataframe containing 10 records referencing Runways that contain all relevant information (from the 102 records containing NO UAS_Distance information)`

In [54]:
rwy10_dec = pd.merge(rwy10, air_dec, on = 'IDENT', how = 'left')

# Only keep necessary columns/fields of interest
rwy10_decDF = pd.DataFrame(rwy10_dec[['REMARKS', 'UASLOCATION', 'IDENT', 'UAS_Location_Runways', 'UAS_Distance', 'Distance_Kilometers', 'RWYDESIGNATOR', 'Runway.Latitude', 'Runway.Longitude', 'DECLINATION']])

### Calculate Bearing Information for Runways

`To do this, you take the number portion of the Runway Designator (in the 'RWYDESIGNATOR' field) and multiply it by 10 (effectively adding a 0 to the end) and add the DECLINATION to this number`

***Example:*** *RWY 32R has designator 32R which indicates a bearing of 320 degrees*

In [55]:
designator = rwy10_decDF['RWYDESIGNATOR']
designator_bearing = []

for i in range(len(rwy10_decDF)):
    designator_bearing.append(int(re.sub("\D","",designator[i]))*10 + rwy10_decDF['DECLINATION'][i])

rwy10_decDF['RWY_BEARING'] = designator_bearing

In [56]:
rwy10_decDF.rename(columns = {'Runway.Latitude': 'Runway_Longitude', 'Runway.Longitude' : 'Runway_Latitude'}, inplace = True)

### Calculate UAS Latitude/Longitude Information using geopy library

In [57]:
uas_lat = []
uas_long = []


for i in range(len(rwy10_decDF)):
    lat_rwy = pd.to_numeric(rwy10_decDF['Runway_Latitude'][i])
    long_rwy = pd.to_numeric(rwy10_decDF['Runway_Longitude'][i])
    b = pd.to_numeric(rwy10_decDF['RWY_BEARING'][i])
    d = pd.to_numeric(rwy10_decDF['Distance_Kilometers'][i])

    origin = geopy.Point(lat_rwy, long_rwy)
    destination = geodesic(kilometers=d).destination(origin,b)

    lat2, lon2, = destination.latitude, destination.longitude

    uas_lat.append(lat2)
    uas_long.append(lon2)

# Append UAS Lat/Long information to DataFrame
rwy10_decDF['UAS_Latitude'] = uas_lat
rwy10_decDF['UAS_Longitude'] = uas_long

In [58]:
rwy10_decDF.drop_duplicates(subset = ['REMARKS']).reset_index(drop=True)

REMARKS UASLOCATION IDENT  \
0  Aircraft observed a UAS at the 2 o'clock posit...         MKE   MKE   
1  Aircraft observed a UAS SW bound at 900 feet o...         LGA   LGA   
2  Aircraft observed a silver UAS on 4 mile final...         JFK   JFK   
3  Aircraft observed a UAS while descending throu...         ACK   ACK   
4  Aircraft observed a UAS while descending throu...         ACK   ACK   
5  Aircraft observed a UAS 800 feet overhead off ...         SGJ   SGJ   
6  Aircraft observed a black and silver UAS from ...         LAX   LAX   
7  Aircraft observed a white UAS off the left sid...         LAX   LAX   
8  Aircraft observed a UAS from the 12 o'clock wh...         EWR   EWR   
9  Aircraft reported a UAS at 1500 ft. while on 3...         ENW   ENW   

  UAS_Location_Runways      UAS_Distance  Distance_Kilometers RWYDESIGNATOR  \
0          ['RWY 19R']  [1.5 mile final]                2.778           19R   
1           ['RWY 22']  [2.5 mile final]                4.630            22   
2          ['RWY 31R']    [4 mile final]                7.408           31R   
3           ['RWY 24']    [2 mile final]                3.704            24   
4           ['RWY 24']    [2 mile final]                3.704            24   
5           ['RWY 13']    [1 mile final]                1.852            13   
6          ['RWY 25L']    [2 mile final]                3.704           25L   
7          ['RWY 25L']    [3 mile final]                5.556           25L   
8          ['RWY 22R']    [3 mile final]                5.556           22R   
9          ['RWY 25R']    [3 mile final]                5.556           25R   

   Runway_Longitude  Runway_Latitude  DECLINATION  RWY_BEARING  UAS_Latitude  \
0        -87.895715        42.941378     -3.97494    186.02506     42.916510   
1        -73.878696        40.775660    -12.75395    207.24605     40.738590   
2        -73.777891        40.652110    -12.77039    297.22961     40.682607   
3        -70.066003        41.251812    -14.42364    225.57636     41.228463   
4        -70.066003        41.251812    -14.42364    225.57636     41.228463   
5        -81.338523        29.959573     -6.83122    123.16878     29.950432   
6       -118.404501        33.935085     11.71212    261.71212     33.930265   
7       -118.404501        33.935085     11.71212    261.71212     33.927851   
8        -74.172492        40.686242    -12.59521    207.40479     40.641820   
9        -87.932913        42.596187     -3.90298    246.09702     42.575905   

   UAS_Longitude  
0     -87.899286  
1     -73.903791  
2     -73.855809  
3     -70.097554  
4     -70.097554  
5     -81.322464  
6    -118.444143  
7    -118.463962  
8     -74.202724  
9     -87.994783

In [59]:
# Export to .csv 
rwy10_decDF.to_csv('rwy10_uasLatLong.csv', index = False)

#### Obtain Dataframe (and corresponding .csv file) for the 2 records for which UAS Lat/Long information was not captured

In [60]:
rwy_unid2 = rwy102_df_MF.iloc[[1, 9]]

In [61]:
rwy_unid2

index                                            REMARKS  \
1    132  Aircraft observed a  silver quadcopter UAS fro...   
9    376  Aircraft observed a white UAS off the left sid...   

  UAS_Location_Runways UASLOCATION IDENT    UAS_Distance  Distance_Kilometers  
1          ['RWY 17R']         AUS   AUS  [7 mile final]               12.964  
9          ['RWY 25L']         LAX   LAX  [3 mile final]                5.556

In [62]:
# Export to .csv 
rwy_unid2.to_csv('rwy2_EXCEPTIONS.csv', index = False)

## Exploring the 90 records for which there is truly no Distance information in the Remark

`These 90 records are a subset of the 102 records initially thought to contain no Distance information regarding the UAS sighting`

In [63]:
rwy_unid90 = rwy102_df[rwy102_df['UAS_Distance'].isnull()].reset_index(drop=True)

In [64]:
rwy_unid90

index                                            REMARKS  \
0      97  Aircraft observed a flight of 2 white UAS on t...   
1     104  Aircraft observed a UAS off the left side whil...   
2     107  Aircraft observed a UAS while NE bound at 900 ...   
3     111  Aircraft observed a UAS while S bound at 1,000...   
4     114  Aircraft observed a UAS over the yop while NE ...   
..    ...                                                ...   
85    589  Aircraft reported a red and white quad-copter ...   
86    590  Aircraft reported a UAS off left side, while o...   
87    593  Aircraft reported a UAS at 4,400 feet while on...   
88    597  Aircraft reported a grey quad-copter UAS off t...   
89    600  VFR Aircraft reported a near mid-air with a UA...   

   UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0            ['RWY 24']         CRQ   CRQ      RWY 24          NaN  
1           ['RWY 23L']         RDU   RDU     RWY 23L          NaN  
2            ['RWY 28']         MYF   MYF      RWY 28          NaN  
3           ['RWY 26R']         LGB   LGB     RWY 26R          NaN  
4            ['RWY 22']         LWB   LWB      RWY 22          NaN  
..                  ...         ...   ...         ...          ...  
85           ['RWY 15']         PNE   PNE      RWY 15          NaN  
86          ['RWY 35L']         DWH   DWH     RWY 35L          NaN  
87          ['RWY 35C']         DFW   DFW     RWY 35C          NaN  
88          ['RWY 27L']         SFB   SFB     RWY 27L          NaN  
89          ['RWY 28R']         BOI   BOI     RWY 28R          NaN  

[90 rows x 7 columns]

In [65]:
# Export to .csv 
rwy_unid90.to_csv('rwy90_unid.csv', index = False)

# Look at the records for which there is NOT a valid IDENT

`There are 302 records (within the 504 unidentified runway remarks) that have valid IDENT/REPORTINGFACILITY Information associated with the Runway mentioned within the Remark.`

In [66]:
rwy302 = rwy_unid504[rwy_unid504['IDENT'].isnull()].reset_index(drop=True)
rwy302

index                                            REMARKS  \
0       95  Aircraft observed a white UAS at 500 feet whil...   
1       96  Aircraft observed a UAS off the right side whi...   
2       98  Aircraft observed a black Quad-copter UAS off ...   
3       99  Aircraft observed a white UAS off the left sid...   
4      100  Aircraft observed a UAS 100 feet below while o...   
..     ...                                                ...   
297    587  Aircraft reported a green quad-copter UAS off ...   
298    592  Aircraft reported a white quad-copter UAS off ...   
299    595  Aircraft reported a black quad-copter UAS appr...   
300    598  Aircraft reported a black quad-copter UAS off ...   
301    599  Aircraft reported a NMAC with a UAS while on a...   

    UAS_Location_Runways UASLOCATION IDENT RWYLOCATION UAS_Distance  
0             ['RWY 9L']         NaN   NaN         NaN           []  
1            ['RWY 24R']         NaN   NaN         NaN      ['5NM']  
2             ['RWY17L']         NaN   NaN         NaN           []  
3             ['RWY 8L']         NaN   NaN         NaN      ['3NM']  
4              ['RWY29']         NaN   NaN         NaN           []  
..                   ...         ...   ...         ...          ...  
297           ['RWY 4L']         NaN   NaN         NaN     ['2 NM']  
298           ['RWY 10']         NaN   NaN         NaN      ['1NM']  
299            ['RWY 7']         NaN   NaN         NaN     ['2 NM']  
300           ['RWY 3R']         NaN   NaN         NaN           []  
301            ['RWY05']         NaN   NaN         NaN     ['6 NM']  

[302 rows x 7 columns]